lib/graphql/batch/execution_strategy.rb

```diff
@@ -1,49 +1,44 @@
 module GraphQL::Batch
   class ExecutionStrategy < GraphQL::Query::SerialExecution
-    class OperationResolution < GraphQL::Query::SerialExecution::OperationResolution
-      def result
-        GraphQL::Batch::Promise.resolve(super).sync
-      ensure
-        GraphQL::Batch::Executor.current.clear
-      end
+    def execute(_, _, _)
+      as_promise_unless_resolved(super).sync
+    ensure
+      GraphQL::Batch::Executor.current.clear
     end
 
-    class SelectionResolution < GraphQL::Query::SerialExecution::SelectionResolution
-      def result
-        as_promise_unless_resolved(super)
-      end
-
-      private
+    private
 
-      def as_promise_unless_resolved(result)
-        all_promises = []
-        each_promise(result) do |obj, key, promise|
-          obj[key] = nil
-          all_promises << promise.then { |value| obj[key] = value }
+    def as_promise_unless_resolved(result)
+      all_promises = []
+      each_promise(result) do |obj, key, promise|
+        obj[key] = nil
+        all_promises << promise.then do |value|
+          obj[key] = value
+          as_promise_unless_resolved(value)
         end
-        return result if all_promises.empty?
-        Promise.all(all_promises).then { result }
       end
+      return result if all_promises.empty?
```

Comment:
Previously the code was unconditionally returning a promise in the `.then` block.  I'm actually surprised that didn't result in infinite recursion.  Relying on that behaviour seems fragile and confusing to read.